### ISSUE TITLE
Improve the CLI

### ISSUE THREAD

**Issue by LuiggiTenorioK** [BSC-ES/autosubmit-api#48]
Right now, the API includes a CLI to facilitate its setup without figuring out how to serve it with gunicorn by using just `autosubmit_api start`. Even so, I think that we can make some improvements to better help the end user. These are the ones that I have in mind:

* Add a subcommand `autosubmit_api stop [port]` to stop all or one (by port) API daemon instance that is running.
* Add a subcommand `autosubmit_api check` to list all the running instances of the API with their respective assigned port. This might be possible by using `ps x | grep autosubmit_api`. Also, it will help to implement `autosubmit_api stop` as well.
* Add a subcommand `autosubmit_api gui` or update the `autosubmit_api start` to serve the GUI. This could help a lot as I found that it might be a problem to map which versions of the GUI are compatible with what versions of the API. Minding more or less how the Cylc GUI (and the [Django admin site](https://docs.djangoproject.com/en/4.2/ref/contrib/admin/#)) works, we can have the API and the GUI as the same package. It looks feasible to do by having the current compatible version of the GUI already compiled along with the API source.

What do you think about these options? @mcastril @kinow

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Nov 20, 2023, 17:19

>Add a subcommand `autosubmit_api stop [port]` to stop all or one (by port) API daemon instance that is running.

Playing devil's advocate here; wouldn't it be easier to have one command to start it, and stop when quitting (CTRL+C, signals, etc.)?

I think when developing you wouldn't need to run it as a daemon. And when we create a Linux daemon service we wouldn't call stop.

>Add a subcommand `autosubmit_api check` to list all the running instances of the API with their respective assigned port. This might be possible by using `ps x | grep autosubmit_api`. Also, it will help to implement `autosubmit_api stop` as well.

I think you would have to provide the PID of the instance if you are going to have multiple processes. I am used to just `ps aux | grep blabla` and then `netstat -tlnp $pid`. The issue is that depending on your user permissions you i) not have access to view the open ports or ii) `netstat` or other tools might not be installed (like in some lightweight containers).

Also, an API process started by the system probably cannot be stopped by your user unless you have the right privileges (i.e. on the Climate DT VM I **think** we would have that scenario, so noone would stop the API, only Henrik&CSC ops via `systemctl` or `service`?).

>Add a subcommand `autosubmit_api gui` or update the `autosubmit_api start` to serve the GUI. This could help a lot as I found that it might be a problem to map which versions of the GUI are compatible with what versions of the API. Minding more or less how the Cylc GUI (and the [Django admin site](https://docs.djangoproject.com/en/4.2/ref/contrib/admin/#)) works, we can have the API and the GUI as the same package. It looks feasible to do by having the current compatible version of the GUI already compiled along with the API source.

I think that could be easier. You install the API and don't have to worry about the GUI version, and you also don't need another process serving the UI. Maybe if we have a command line option like `--port` (not sure if we don't already have it) then we could run multiple instances of the API by simply changing the port.

**Comment by LuiggiTenorioK** [comment]
> Playing devil's advocate here; wouldn't it be easier to have one command to start it, and stop when quitting (CTRL+C, signals, etc.)?

Yes, that's how it actually works. The idea was just to extend it to stop when creating daemon processes.

> I think you would have to provide the PID of the instance if you are going to have multiple processes. I am used to just `ps aux | grep blabla` and then `netstat -tlnp $pid`. The issue is that depending on your user permissions you i) not have access to view the open ports or ii) `netstat` or other tools might not be installed (like in some lightweight containers).

I was aware that `netstat` may not be installed, so I was limiting dependencies by using just `ps` and parsing the full command to get the port as it is defined by the `-b` argument in `autosubmit_api start`. 

> Also, an API process started by the system probably cannot be stopped by your user unless you have the right privileges (i.e. on the Climate DT VM I **think** we would have that scenario, so noone would stop the API, only Henrik&CSC ops via `systemctl` or `service`?).

Yeah, it already happened. I couldn't stop the instance that Rohan was running and had to ask him to do the procedures described in `autosubmit_api check` and `autosubmit_api stop`.

>I think that could be easier. You install the API and don't have to worry about the GUI version, and you also don't need another process serving the UI. Maybe if we have a command line option like `--port` (not sure if we don't already have it) then we could run multiple instances of the API by simply changing the port.

Sure! I think this is the best feature to implement in the CLI right now to also ensure the portability of the GUI ASAP. Even so, I need to be sure that this idea can work correctly along with `Flask` and `gunicorn`.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Nov 30, 2023, 09:51

I still find the `stop` useful in case it is left in a closed terminal using nohup. What do you think?

So, do you think it would be better to deploy it as a service? What do you think?

About `we can have the API and the GUI as the same package` I am not sure if I got it well. We need to discuss. There can be users that install the API to be consumed by their own applications, like in EDITO. I would keep them totally separated.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @mcastril on Nov 30, 2023, 09:52

Also I am not sure that I understood the compatibility issue. The API should always preserve backwards compatibility until we change the major version. We add endpoints but we shouldn't remove them.

**Comment by LuiggiTenorioK** [comment]
In GitLab by @kinow on Nov 30, 2023, 09:56

They would be kept separately. When you installed `autosubmit_api` it would come with the Flask app serving the API. Should you want, you could use the GUI that comes bundled as Flask static files. But if you prefer to use just the API you can. If you want to use another GUI, you can do that as well.

The advantage of this approach is that you save the user from having to run two distinct services and have to handle the UI to API configuration, CORS, reverse proxy, etc.

**Comment by LuiggiTenorioK** [comment]
Of course, by compatibility, I pointed to the extended features that new versions of API will include that old versions of the GUI will not handle.

Also, as @kinow mentioned, it facilitates the user to avoid doing an additional installation if they want, and not limit them to installing any GUI version they want.

**Comment by LuiggiTenorioK** [comment]
mentioned in commit 4a6572db6787eac8da7b94c845a387143f03f717

**Comment by ntorqulu** [comment]
1. `--port/ --bind` multi-instance already done in pr #169 
2. `autosubmit_api check` and `autosubmit_api stop [port]`: interesting to have. Write one pidfile per instance. Could make it rely on pidfiles + fallback to `ps` for backward-compatibility for old instances that did not generate a pidfile.
3. Keep API and GUI separated
4. Update `README.md` to include how to start the server without needing authentication (for testing purposes only, never in production)
